### ISSUE TITLE
V0.3.6 cannot ‘-normaliseconf’ if ‘Listen’ isn't an array

### ISSUE THREAD

**Issue by kseistrup** [yggdrasil-network/yggdrasil-go#483]
I had problems upgrading from v0.3.5 to v0.3.6, and I didn't find the error messages helpful. Yggdrasil didn't seem to like the `/etc/yggdrasil.conf` that I used with v0.3.5, so I tried to `-normaliseconf` first:

```sh
# yggdrasil below is v0.3.6
$ yggdrasil -useconf -normaliseconf < yggdrasil.conf
panic: 13 error(s) decoding:

* 'Listen[0]' expected type 'string', got unconvertible type 'uint8'
* 'Listen[10]' expected type 'string', got unconvertible type 'uint8'
* 'Listen[11]' expected type 'string', got unconvertible type 'uint8'
* 'Listen[12]' expected type 'string', got unconvertible type 'uint8'
* 'Listen[1]' expected type 'string', got unconvertible type 'uint8'
* 'Listen[2]' expected type 'string', got unconvertible type 'uint8'
* 'Listen[3]' expected type 'string', got unconvertible type 'uint8'
* 'Listen[4]' expected type 'string', got unconvertible type 'uint8'
* 'Listen[5]' expected type 'string', got unconvertible type 'uint8'
* 'Listen[6]' expected type 'string', got unconvertible type 'uint8'
* 'Listen[7]' expected type 'string', got unconvertible type 'uint8'
* 'Listen[8]' expected type 'string', got unconvertible type 'uint8'
* 'Listen[9]' expected type 'string', got unconvertible type 'uint8'

goroutine 1 [running]:
main.readConfig(0xc00009e6b9, 0xc0000889e0, 0xc00009e6ba, 0x2)
	…/yggdrasil-go/cmd/yggdrasil/main.go:85 +0x459
main.main()
	…/yggdrasil-go/cmd/yggdrasil/main.go:134 +0x1e73

# yggdrasil was not being ehlpful here, was it.
```

I got `yggdrasil` up and running again by -normaliseconf'ing first with v0.3.5, then with v0.3.6.

Yggdrasil has been compiled with “go version go1.12.7 linux/amd64” (go-pie, actually) on ArchLinux.

**Comment by ChronosX88** [comment]
Just change in config Listen parameter:
```
Listen: ["tcp://[::]:0"]
```
and it will work!

**Comment by kseistrup** [comment]
> Just change

True, but (1) that shouldn't be necessary, and (2) that requires the user to know why the current value is rejected. Also, »expected type 'string', got unconvertible type 'uint8'« is an extremely unhelpful error message. And what's with the strange sorting of the `Listen[]` lines?

At least an error message like “`Listen` requires an array type of parameter but found a string” would have helped somewhat.

**Comment by ChronosX88** [comment]
Yes, I agree with you :-)

**Comment by neilalexander** [comment]
This is a bit of an unfortunate side-effect of how we read config into the node. The config is defined in the `NodeConfig` struct, and we use `hjson` and `mapstructure` to take in the HJSON config and map it onto an instance of that struct. The error is indeed unhelpful.

It might be possible to do an intermediate step somewhere to check the value we've read in from HJSON before we use `mapstructure` - I can take a look.

**Comment by neilalexander** [comment]
This has been fixed by #492, which should hopefully make life a bit easier for anyone jumping from v0.3.5 to v0.3.7 when we release it.
